### PR TITLE
Upgrade github.com/mostynb/go-grpc-compression and switch to nonclobbering imports

### PR DESCRIPTION
**Pull request by mostynb**
**Description:** Switch to grpc encoders that won't override the consumer's previously registered codecs.

Some consumers of this library have registered grpc compressors, and the init() order can cause their compressors to be clobbered by opentelemetry-collector's use of github.com/mostynb/go-grpc-compression, which follows the pattern from the grpc-go library (overwrite any previously registered encoders with the same name).

Version 1.2.0 of github.com/mostynb/go-grpc-compression added so-called "nonclobbering" imports, which only register the codec if there is not already a codec registered with that name. This PR switches to use those.

**Link to tracking Issue:** #7920

**Testing:** tests have been added to the upstream library.

### CODE REVIEW

**Review by codeboten — APPROVED** [review]
Thanks @mostynb

**Review by codeboten — COMMENTED** [review]
@mostynb can you add a [changelog entry](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#adding-a-changelog-entry) to capture this change?

I suspect it will mostly be a non event for most users, but in case anyone is surprised by their compression options changing, they can reference the changelog to see what changed.

**Comment by mostynb** [comment]
> @mostynb can you add a [changelog entry](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#adding-a-changelog-entry) to capture this change?
> 
> I suspect it will mostly be a non event for most users, but in case anyone is surprised by their compression options changing, they can reference the changelog to see what changed.

I gave it a try- I hope the formatting is OK.

**Review by codeboten — APPROVED** [review]
Thanks for rebasing @mostynb!

File: config/configgrpc/go.sum [hunk 1]
```diff
@@ -291,7 +291,6 @@ github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd/go.mod h1:6dJ
 github.com/modern-go/reflect2 v0.0.0-20180701023420-4b7aa43c6742/go.mod h1:bx2lNnkwVCuqBIxFjflWJWanXIb3RllmbCylyMrvgv0=
 github.com/modern-go/reflect2 v1.0.1/go.mod h1:bx2lNnkwVCuqBIxFjflWJWanXIb3RllmbCylyMrvgv0=
 github.com/modern-go/reflect2 v1.0.2 h1:xBagoLtFs94CBntxluKeaWgTMpvLxC4ur3nMaC9Gz0M=
-github.com/modern-go/reflect2 v1.0.2/go.mod h1:yWuevngMOJpCy52FWWMvUC8ws7m/LJsjYzDa0/r8luk=
```
Comment (mostynb) [inline config/configgrpc/go.sum:294]: oops, this should not have been removed

**Pull request merged**